internal/auditd/sessiontracker.go

```diff
@@ -128,102 +127,115 @@ func (o *sessionTracker) auditdEvent(event *aucoalesce.Event) error {
 		return nil
 	}
 
-	var debugLogger *zap.SugaredLogger
-	if logger.Level().Enabled(zap.DebugLevel) {
-		debugLogger = logger.With(
-			"auditEvent", *event,
-			"auditEventType", event.Type.String(),
-			"auditSessionID", event.Session)
-		debugLogger.Debugln("new audit event")
-	}
+	debugLogger := logger.With(
+		"auditEvent", *event,
+		"auditEventType", event.Type.String(),
+		"auditSessionID", event.Session)
+	debugLogger.Debugln("new audit event")
 
-	u, hasSession := o.sessIDsToUsers.Load(event.Session)
-	//nolint:nestif // Refactor later.
-	if hasSession {
-		// Either write the audit event to the event writer
-		// or cache it for later.
-
-		if debugLogger != nil {
-			debugLogger.With(
-				"auditSessionStartTime", u.added,
-				"numCachedAuditEvents", len(u.cached),
-				"hasRUL", u.hasRUL).
-				Debugln("found existing audit session for audit event")
-		}
-
-		if u.hasRUL {
-			// It looks like AUDIT_CRED_DISP indicates the
-			// canonical end of a user session - but, there is
-			// also AUDIT_USER_END, which occurs just before.
-			if event.Type == auparse.AUDIT_CRED_DISP {
-				defer o.sessIDsToUsers.Delete(event.Session)
-			}
+	if o.sessIDsToUsers.Has(event.Session) {
+		return o.auditEventWithSession(event, debugLogger)
+	}
 
-			err := u.writeAndClearCache(o.eventWriter)
-			if err != nil {
-				return &sessionTrackerError{
-					auditEventFail: true,
-					message: fmt.Sprintf("failed to write cached events for user '%s' - %s",
-						u.login.CredUserID, err),
-					inner: err,
-				}
-			}
+	return o.auditEventWithoutSession(event, debugLogger)
+}
 
-			err = o.eventWriter.Write(u.toAuditEvent(event))
-			if err != nil {
-				return &sessionTrackerError{
-					auditEventFail: true,
-					message:        err.Error(),
-					inner:          err,
-				}
-			}
+// auditEventWithSession handles an audit event that is associated with
+// an audit session. If the user object associated with the session
+// already has remote user login information, the event is written to the
+// audit event writer. Otherwise, the event is cached in the user object.
+//
+// Note that this is all done within a lock on the user object. So it's
+// thread-safe. However, recall the WithLockedValueDo method holds
+// the lock for the duration of the callback. So this method should
+// return quickly and shouldn't call any other locking methods on the
+// sessionTracker.
+func (o *sessionTracker) auditEventWithSession(event *aucoalesce.Event, debugLogger *zap.SugaredLogger) error {
+	return o.sessIDsToUsers.WithLockedValueDo(event.Session, func(u *user) error {
```

Comment:
Definitely makes thread safe :) 

So this got me thinking about processing multiple sessions at one time. the session tracker is tracking multiple sessions and adds an event to the `*user`s cached events. The only thing we really need to lock for `user` is when we are modifying the user. i.e. appending to cache, writing and clearing the cache etc...
 
 An idea that might work is to keep the `sessiontracker` really simple. All it would do is add new events to a users cached slice and we would lock around this operation. We could create another struct called `sessionworker` that does all the processing for a particular session. that way we synchronize the incoming logs but can do all the processing asynchronously. This follows a `manager` to many `employees` type of pattern 
 
 whenever the session ends the `sessionworker` go routine would signal to the `sessiontracker` to remove that session from the list of sessions. 
 
 just a thought. The thread safe code written solves the problem i had mentioned in the previous PR. :)